File: lightroom_tagger/core/fallback.py
"""FallbackDispatcher — tries selected provider, cascades through fallback
order on retryable failures."""

from __future__ import annotations

from typing import Any, Callable

from lightroom_tagger.core.cancel_scope import resolve_cancel_check
from lightroom_tagger.core.error_policy import (
    ConsecutiveAbortTracker,
    ErrorPolicy,
    NoOpErrorPolicy,
)
from lightroom_tagger.core.exceptions import (
    RETRYABLE_ERRORS,
    ConnectionError,
    InvalidRequestError,
    ModelUnavailableError,
    ProviderError,
    RateLimitError,
)
from lightroom_tagger.core.provider_registry import ProviderRegistry
from lightroom_tagger.core.retry import CancelledRetryError, retry_with_backoff

LogCallback = Callable[[str, str], None] | None


def _log_cascade(
    log_callback: LogCallback,
    operation: str,
    pid: str,
    mid: str,
    exc: Exception,
    attempts: list[tuple[str, str]],
    index: int,
) -> None:
    """Emit a single, uniform log line when a provider attempt falls through."""
    if not log_callback:
        return
    remaining = attempts[index + 1:]
    next_label = f"{remaining[0][0]}/{remaining[0][1]}" if remaining else "none"
    log_callback(
        "warning",
        f"[{operation}] {pid}/{mid} failed ({type(exc).__name__}), "
        f"fallback -> {next_label}",
    )


class FallbackDispatcher:
    """Wraps provider calls with retry + automatic cascade."""

    def __init__(
        self,
        registry: ProviderRegistry,
        error_policy: ErrorPolicy | None = None,
    ):
        self._registry = registry
        self._error_policy = (
            error_policy if error_policy is not None else NoOpErrorPolicy()
        )

    @property
    def error_policy(self) -> ErrorPolicy:
        return self._error_policy

    def call_with_fallback(
        self,
        operation: str,
        fn_factory: Callable,
        provider_id: str,
        model: str,
        log_callback: LogCallback = None,
        cancel_check: Callable[[], bool] | None = None,
        abort_tracker: ConsecutiveAbortTracker | None = None,
    ) -> tuple[Any, str, str]:
        """Execute *fn_factory(client, model)* with retry and fallback.

        Tries the requested model first, then all other models in the same
        provider (in configured order), then moves on to each fallback provider.

        Parameters
        ----------
        operation:
            Label for logging (``"compare"`` or ``"describe"``).
        fn_factory:
            ``(client, model) -> callable`` — returns a zero-arg callable that
            performs the actual API call.
        provider_id:
            Primary provider to try first.
        model:
            Model to use with the primary provider.
        log_callback:
            Optional ``(level, message)`` logger.

        Returns
        -------
        ``(result, actual_provider_id, actual_model)``
        """
        attempts, empty_fallbacks = self._build_attempts(provider_id, model)
        if not attempts:
            raise ProviderError("No available providers for operation")
        last_error: ProviderError | None = None

        cancel_check = resolve_cancel_check(cancel_check)

        for index, (pid, mid) in enumerate(attempts):
            # Cancel between providers: without this, a cancel during the
            # primary provider's retries would still trigger the full
            # fallback cascade (another 3-4 × 42s of work) before the
            # worker checked in again.
            if cancel_check is not None and cancel_check():
                raise CancelledRetryError("cancel requested before fallback attempt")

            if abort_tracker is not None and abort_tracker.rate_limit_abort_reached:
                raise RateLimitError("consecutive rate-limit abort threshold reached")

            client = self._registry.get_client(pid)
            retry_config = self._registry.get_retry_config(pid)
            fn = fn_factory(client, mid)

            try:
                result = retry_with_backoff(
                    fn,
                    retry_config,
                    log_callback=log_callback,
                    cancel_check=cancel_check,
                )
                if abort_tracker is not None:
                    abort_tracker.record_success()
                return result, pid, mid
            except CancelledRetryError:
                # Surface cancellation directly — do NOT fall through to
                # the next provider. ``raise`` keeps the original frame.
                raise
            except InvalidRequestError as exc:
                if abort_tracker is not None:
                    abort_tracker.record_fatal()
                raise
            except tuple(RETRYABLE_ERRORS) as exc:
                last_error = exc  # type: ignore[assignment]
                _log_cascade(log_callback, operation, pid, mid, exc, attempts, index)
            except ConnectionError as exc:
                # ``ConnectionError`` is NOT_RETRYABLE (see commit 5b0763a —
                # "connection refused / DNS failure is permanent for that
                # provider") but it IS a provider-specific failure that
                # should cascade to the next available provider instead of
                # surfacing to the caller. Other NOT_RETRYABLE errors
                # (``AuthenticationError``, ``InvalidRequestError``) are
                # global and propagate below.
                last_error = exc
                _log_cascade(log_callback, operation, pid, mid, exc, attempts, index)

        # All attempts exhausted. If any advertised fallback provider had no
        # vision models we surface that as the *actionable* error — the
        # caller can configure or enable a model — instead of whatever
        # transient error the last working provider happened to throw.
        if empty_fallbacks:
            raise ModelUnavailableError(
                f"No models available for fallback provider(s): "
                f"{', '.join(empty_fallbacks)}"
            )
        if abort_tracker is not None and last_error is not None:
            abort_tracker.record_dispatch_outcome(last_error)
        raise last_error  # type: ignore[misc]

    def _build_attempts(
        self, primary_id: str, primary_model: str
    ) -> tuple[list[tuple[str, str]], list[str]]:
        """Return ordered list of (provider_id, model_id) to try.

        Starts with all vision models in the primary provider (requested model
        first, then the rest in configured order), then the first vision model
        from each fallback provider in fallback order.

        Also returns the list of fallback provider IDs that advertised no
        vision models — used by ``call_with_fallback`` to surface a
        ``ModelUnavailableError`` with an actionable message when the
        cascade exhausts.
        """
        available_ids = {
            entry["id"]
            for entry in self._registry.list_providers()
            if entry.get("available")
        }

        attempts: list[tuple[str, str]] = []
        empty_fallbacks: list[str] = []

        # Primary provider: requested model first, then remaining vision models
        if primary_id in available_ids:
            all_models = self._registry.list_models(primary_id)
            vision_models = [m["id"] for m in all_models if m.get("vision")]
            # Requested model goes first (even if not in the list)
            ordered = [primary_model] + [m for m in vision_models if m != primary_model]
            for mid in ordered:
                attempts.append((primary_id, mid))

        # Fallback providers: first vision model each
        for pid in self._registry.fallback_order:
            if pid == primary_id or pid not in available_ids:
                continue
            all_models = self._registry.list_models(pid)
            vision_models = [m["id"] for m in all_models if m.get("vision")]
            if vision_models:
                attempts.append((pid, vision_models[0]))
            else:
                empty_fallbacks.append(pid)

        return attempts, empty_fallbacks

File: apps/visualizer/backend/src/providers/fallback.ts
/**
 * Try the selected provider, then cascade through the fallback order.
 *
 * The attempt list is wider than it looks: every vision model in the primary
 * provider (requested one first), then the first vision model from each fallback
 * provider. A local Ollama that has the model unloaded should not send the whole
 * batch to a paid API, and a paid API that is rate-limiting should not stall the
 * batch when a local model is available.
 */
import {
  InvalidRequestError,
  isRetryableError,
  ModelUnavailableError,
  ProviderConnectionError,
  ProviderError,
} from './errors.js';
import type { ProviderRegistry } from './registry.js';
import type { ProviderClient } from './vision-client.js';
import { CancelledRetryError, retryWithBackoff, type CancelCheck, type LogCallback } from './retry.js';

/** `(client, model) -> () => Promise<result>` */
export type FnFactory<T> = (client: ProviderClient, model: string) => () => Promise<T>;

/** One uniform line when a provider attempt falls through. */
function logCascade(
  logCallback: LogCallback,
  operation: string,
  pid: string,
  mid: string,
  e: unknown,
  attempts: readonly [string, string][],
  index: number,
): void {
  if (!logCallback) return;
  const next = attempts[index + 1];
  const nextLabel = next ? `${next[0]}/${next[1]}` : 'none';
  logCallback(
    'warning',
    `[${operation}] ${pid}/${mid} failed (${(e as Error).name}), fallback -> ${nextLabel}`,
  );
}

export class FallbackDispatcher {
  private readonly registry: ProviderRegistry;

  constructor(registry: ProviderRegistry) {
    this.registry = registry;
  }

  /**
   * Execute `fnFactory(client, model)` with retry and fallback.
   *
   * Returns the result along with the provider and model that actually served
   * it, which the caller records — the label written into `model_used` has to be
   * what ran, not what was asked for.
   */
  async callWithFallback<T>(
    opts: {
      operation: string;
      fnFactory: FnFactory<T>;
      providerId: string;
      model: string;
      logCallback?: LogCallback;
      cancelCheck?: CancelCheck;
    },
  ): Promise<{ result: T; providerId: string; model: string }> {
    const { attempts, emptyFallbacks } = await this.buildAttempts(opts.providerId, opts.model);
    if (attempts.length === 0) throw new ProviderError('No available providers for operation');

    const logCallback = opts.logCallback ?? null;
    const cancelCheck = opts.cancelCheck ?? null;
    let lastError: unknown = null;

    for (const [index, [pid, mid]] of attempts.entries()) {
      // Checked between providers: without it, a cancel during the primary
      // provider's retries would still drag the whole cascade through another
      // three or four rounds of backoff before the worker checked in again.
      if (cancelCheck !== null && cancelCheck()) {
        throw new CancelledRetryError('cancel requested before fallback attempt');
      }

      const client = this.registry.getClient(pid);
      const retryConfig = this.registry.getRetryConfig(pid);
      const fn = opts.fnFactory(client, mid);

      try {
        const result = await retryWithBackoff(fn, retryConfig, { logCallback, cancelCheck });
        return { result, providerId: pid, model: mid };
      } catch (e) {
        // A cancel is surfaced directly — it must not fall through to the next
        // provider, which would be doing work the user asked us to stop.
        if (e instanceof CancelledRetryError) throw e;

        if (e instanceof InvalidRequestError) throw e;

        if (isRetryableError(e) || e instanceof ProviderConnectionError) {
          // `ProviderConnectionError` is not retryable — a refused connection is
          // permanent for that provider — but it IS provider-specific, so it
          // cascades. `AuthenticationError` and `InvalidRequestError` are global
          // and propagate instead.
          lastError = e;
          logCascade(logCallback, opts.operation, pid, mid, e, attempts, index);
          continue;
        }
        throw e;
      }
    }

    // Exhausted. A fallback provider advertising no vision models is the
    // *actionable* failure — the user can configure one — so it wins over
    // whatever transient error the last working provider happened to throw.
    if (emptyFallbacks.length > 0) {
      throw new ModelUnavailableError(
        `No models available for fallback provider(s): ${emptyFallbacks.join(', ')}`,
      );
    }
    throw lastError;
  }

  /**
   * The ordered `(provider, model)` attempts, plus the fallback providers that
   * turned out to have no vision models.
   */
  private async buildAttempts(
    primaryId: string,
    primaryModel: string,
  ): Promise<{ attempts: [string, string][]; emptyFallbacks: string[] }> {
    const availableIds = new Set(
      (await this.registry.listProviders()).filter((p) => p.available).map((p) => p.id),
    );

    const attempts: [string, string][] = [];
    const emptyFallbacks: string[] = [];

    if (availableIds.has(primaryId)) {
      const visionModels = (await this.registry.listModels(primaryId))
        .filter((m) => m.vision)
        .map((m) => m.id);
      // The requested model goes first even when it is not in the list — the
      // user may have typed a model the registry has not discovered yet.
      const ordered = [primaryModel, ...visionModels.filter((m) => m !== primaryModel)];
      for (const mid of ordered) attempts.push([primaryId, mid]);
    }

    for (const pid of this.registry.fallbackOrder) {
      if (pid === primaryId || !availableIds.has(pid)) continue;
      const visionModels = (await this.registry.listModels(pid))
        .filter((m) => m.vision)
        .map((m) => m.id);
      if (visionModels.length > 0) attempts.push([pid, visionModels[0]!]);
      else emptyFallbacks.push(pid);
    }

    return { attempts, emptyFallbacks };
  }
}
